pkg/agent/proxy/endpointslicecache.go

```diff
@@ -0,0 +1,365 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Original file https://raw.githubusercontent.com/kubernetes/kubernetes/0c0d4fea8dd6bdcd16b9e1d35da3f7d209341a6f/pkg/proxy/endpointslicecache.go
+// Remove makeEndpointInfo and recorder in fields.
+// Remove unused standardEndpointInfo.
+// Remove unneeded sort.Sort in endpointsMapFromEndpointInfo.
+// Update import paths.
+
+package proxy
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+	apimachinerytypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	utilnet "k8s.io/utils/net"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/proxy/types"
+	"github.com/vmware-tanzu/antrea/third_party/proxy"
+)
+
+// EndpointSliceCache is used as a cache of EndpointSlice information.
+type EndpointSliceCache struct {
+	// lock protects trackerByServiceMap.
+	lock sync.Mutex
+
+	// trackerByServiceMap is the basis of this cache. It contains endpoint
+	// slice trackers grouped by service name and endpoint slice name. The first
+	// key represents a namespaced service name while the second key represents
+	// an endpoint slice name. Since endpoints can move between slices, we
+	// require slice specific caching to prevent endpoints being removed from
+	// the cache when they may have just moved to a different slice.
+	trackerByServiceMap map[apimachinerytypes.NamespacedName]*endpointSliceTracker
+
+	hostname   string
+	isIPv6Mode *bool
+}
+
+// endpointSliceTracker keeps track of EndpointSlices as they have been applied
+// by a proxier along with any pending EndpointSlices that have been updated
+// in this cache but not yet applied by a proxier.
+type endpointSliceTracker struct {
+	applied endpointSliceInfoByName
+	pending endpointSliceInfoByName
+}
+
+// endpointSliceInfoByName groups endpointSliceInfo by the names of the
+// corresponding EndpointSlices.
+type endpointSliceInfoByName map[string]*endpointSliceInfo
+
+// endpointSliceInfo contains just the attributes kube-proxy cares about.
+// Used for caching. Intentionally small to limit memory util.
+type endpointSliceInfo struct {
+	Ports     []discovery.EndpointPort
+	Endpoints []*endpointInfo
+	Remove    bool
+}
+
+// endpointInfo contains just the attributes kube-proxy cares about.
+// Used for caching. Intentionally small to limit memory util.
+// Addresses and Topology are copied from EndpointSlice Endpoints.
+type endpointInfo struct {
+	Addresses []string
+	Topology  map[string]string
+}
+
+// spToEndpointMap stores groups Endpoint objects by ServicePortName and
+// EndpointSlice name.
+type spToEndpointMap map[proxy.ServicePortName]map[string]proxy.Endpoint
+
+// NewEndpointSliceCache initializes an EndpointSliceCache.
+func NewEndpointSliceCache(hostname string, isIPv6Mode *bool) *EndpointSliceCache {
```

Comment:
Do we really have a nil case for `isIPv6Mode` and what does it represent? I didn't find it in the code.
It should just use bool to avoid confusion.

Reply:
OK, using bool also works.